src/Anonymization/Anonymizer/Core/DateAnonymizer.php

```diff
@@ -0,0 +1,197 @@
+                }
+            }
+
+            $this->anonmizeWithDelta($update, $format, $delta);
+
+            return;
+        }
+
+        throw new \InvalidArgumentException("Providing either the 'delta' option, or both 'min' and 'max' options is required.");
+    }
+
+    private function anonmizeWithDateRange(Update $update, string $format, \DateTimeImmutable $min, \DateTimeImmutable $max): void
+    {
+        $diff = $max->diff($min, true);
+
+        if ('date' === $format) {
+            // Compute a diff in number of days.
+            $unit = 'day';
+            $delta = $diff->d + $diff->m * 30 + $diff->y * 360;
+        } else if (68 < $diff->y) {
+            // We hit UNIX timestamp maximum integer limit, and may cause
+            // int overflow or other kind of crashes server side.
+            $unit = 'hour';
+            $delta = $diff->h + $diff->d * 24 + $diff->m * 720 + $diff->y * 8640;
+        } else {
+            $unit = 'second';
+            $delta = $diff->s + $diff->i * 60 + $diff->h * 3600 + $diff->d * 86400 + $diff->m * 2592000 + $diff->y * 31104000;
+        }
+
+        // Cut in half to compute middle date.
+        $delta /= 2;
+        $middleDate = $min->add(\DateInterval::createFromDateString(\sprintf("%d %s", $delta, $unit)));
+
+        $this->anonymizeWithDeltaAndReferenceDate($update, $format, $middleDate, $delta, $unit);
+    }
+
+    private function anonmizeWithDelta(Update $update, string $format, \DateInterval $delta): void
+    {
+        // @todo I wish for a better alternative...
+        // query-builder can deal with \DateInterval by- itself, but we are
+        // randomizing values here, so we need to be able to apply a single
+        // figure random delta, in order to be able to use SQL random at the
+        // right place, otherwise the algorithm would be very complex.. 
+        if ('date' !== $format && $delta->s) {
+            // Please, never use seconds...
+            $delta = $delta->s + $delta->i * 60 + $delta->h * 3600 + $delta->d * 86400 + $delta->m * 2592000 + $delta->y * 31104000;
+            $unit = 'second';
+        } else if ('date' !== $format && $delta->i) {
+            $delta = $delta->i + $delta->h * 60 + $delta->d * 1440 + $delta->m * 43200 + $delta->y * 518400;
+            $unit = 'minute';
+        } else if ('date' !== $format && $delta->h) {
+            $delta = $delta->h + $delta->d * 24 + $delta->m * 720 + $delta->y * 8640;
+            $unit = 'hour';
+        } else if ($delta->d) {
+            $delta = $delta->d + $delta->m * 30 + $delta->y * 360;
+            $unit = 'day';
+        } else if ($delta->m) {
+            $delta = $delta->m + $delta->y * 12;
+            $unit = 'month';
+        } else if ($delta->y) {
+            $delta = $delta->y;
+            $unit = 'year';
+        } else {
+            throw new \InvalidArgumentException("'delta' option interval is empty.");
+        }
```

Comment:
Since we discussed about that piece of code IRL, I think it could be nice to have some comments to explain choices you did such as considering all month being 30 days long.

Reply:
OK for me :slightly_smiling_face: